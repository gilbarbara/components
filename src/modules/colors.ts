import { rangeLimit } from '@gilbarbara/helpers';
import { hex2hsl, hsl2hex, textColor } from 'colorizr';

export function getColorScale(color: string) {
  const { h, l, s } = hex2hsl(color);

  const lightest = hsl2hex({ h, s, l: rangeLimit(l * 1.9, 92, 96) });
  const lighter = hsl2hex({ h, s, l: rangeLimit(l * 1.6, 82, 92) });
  const light = hsl2hex({ h, s, l: rangeLimit(l * 1.4, 64, 72) });
  const dark = hsl2hex({ h, s, l: rangeLimit(l * 0.7, 32) });
  const darker = hsl2hex({ h, s, l: rangeLimit(l * 0.45, 16) });
  const darkest = hsl2hex({ h, s, l: rangeLimit(l * 0.3, 8) });

  return {
    lightest: { bg: lightest, color: textColor(lightest) },
    lighter: { bg: lighter, color: textColor(lighter) },
    light: { bg: light, color: textColor(light) },
    mid: { bg: color, color: textColor(color) },
    dark: { bg: dark, color: textColor(dark) },
    darker: { bg: darker, color: textColor(darker) },
    darkest: { bg: darkest, color: textColor(darkest) },
  };
}

export function getGrayScale(
  lightest: string,
  lighter: string,
  light: string,
  mid: string,
  dark: string,
  darker: string,
  darkest: string,
) {
  return {
    lightest: { bg: lightest, color: textColor(lightest) },
    lighter: { bg: lighter, color: textColor(lighter) },
    light: { bg: light, color: textColor(light) },
    mid: { bg: mid, color: textColor(mid) },
    dark: { bg: dark, color: textColor(dark) },
    darker: { bg: darker, color: textColor(darker) },
    darkest: { bg: darkest, color: textColor(darkest) },
  };
}
