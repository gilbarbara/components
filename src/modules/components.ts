import { white } from './theme';

/**
 * Components
 */
export const badge = {
  borderRadius: 10,
  fontSize: 12,
  fontWeight: 700,
  lineHeight: 18,
  padding: 5,
  size: 20,
  sizeDot: 10,
};

export const button = {
  borderRadius: {
    sm: '8px',
    md: '8px',
    lg: '8px',
  },
  dimension: {
    sm: '36px',
    md: '48px',
    lg: '56px',
  },
  fontSize: {
    sm: '14px',
    md: '16px',
    lg: '18px',
  },
  fontWeight: 700,
  lineHeight: {
    sm: 1.2,
    md: 1.3,
    lg: 1.4,
  },
  padding: {
    sm: ['8px', '16px'],
    md: ['10px', '20px'],
    lg: ['12px', '24px'],
  },
};

export const list = {
  borderRadius: 0,
  padding: {
    sm: 8,
    md: 12,
    lg: 16,
  },
};

export const table = {
  borderColors: { primary: '#ccc', secondary: '#32383e' },
  captionColor: '#ccc',
  captionPadding: 6,
  colors: { primary: white, secondary: '#000' },
  headColors: { light: '#f4f4f4', dark: '#000' },
  padding: {
    sm: 6,
    md: 12,
    lg: 18,
  },
  stripedColors: { primary: '#ededed', secondary: '#262626' },
};
