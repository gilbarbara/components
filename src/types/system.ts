import { HTMLAttributes, ReactNode } from 'react';
import { StringOrNumber } from '@gilbarbara/types';
import { StandardLonghandProperties, StandardShorthandProperties } from 'csstype';

import {
  Alignment,
  BorderItem,
  BorderItemSide,
  HeadingSizes,
  Orientation,
  Sizes,
  StringLiteral,
  TextSizes,
  Variant,
} from './common';
import { ButtonSize, ColorVariantTones, Radius, Shadow, Spacing, Theme } from './theme';

export type Gap = Spacing | StringLiteral | number;

export type SpacingAuto = SpacingOrZero | 'auto';
export type SpacingOrZero = Spacing | 0;

export type WithHTMLAttributes<T = HTMLDivElement> = Pick<
  HTMLAttributes<T>,
  'className' | 'id' | 'style' | 'tabIndex' | 'title'
>;

export interface WithAccent<T = ColorVariantTones> {
  /**
   * Component accent color
   * @default primary
   */
  accent?: T;
}

export interface WithAlign {
  /**
   * Text alignment
   */
  align?: Alignment;
}

export interface WithAnimationDuration {
  /**
   * The duration of the animation when the content is entering in milliseconds.
   * @default 300
   */
  animationEnterDuration?: number;
  /**
   * The duration of the animation when the content is exiting in milliseconds.
   * @default 300
   */
  animationExitDuration?: number;
}

export interface WithAnimationType extends WithAnimationDuration {
  /**
   * The type of animation to use for the transition.
   * - grow: Content expands from a small size (only works with 'down' and 'right' placements).
   * - scale: Content scales from a small size.
   * - slide: Content translates into view.
   *
   * @default slide
   */
  animationType?: 'grow' | 'scale' | 'slide';
}

export interface WithBlock {
  /**
   * Use the parent full width
   * @default false
   */
  block?: boolean;
}

export interface WithBorder {
  border?: boolean | BorderItemSide | BorderItem | BorderItem[];
}

export interface WithBorderless {
  /** @default false */
  borderless?: boolean;
}

export interface WithBusy {
  /**
   * Add an animated icon
   * @default false
   */
  busy?: boolean;
}

export interface WithButtonSize {
  /**
   * Button size
   * @default md
   */
  size?: ButtonSize;
}

export interface WithChildren {
  /** Required */
  children: ReactNode;
}

export interface WithChildrenOptional {
  /** Optional */
  children?: ReactNode;
}

export interface WithColors {
  /**
   * Component background color
   */
  bg?: ColorVariantTones;
  /**
   * Component color
   */
  color?: ColorVariantTones;
}

export interface WithColorsDefaultBg extends WithColors {
  /** @default primary */
  bg?: ColorVariantTones;
}

export interface WithColorsDefaultColor extends WithColors {
  /** @default primary */
  color?: ColorVariantTones;
}

export interface WithComponentSize<T = Sizes> {
  /**
   * Component size
   * @default md
   */
  size?: T;
}

export interface WithDimension {
  aspectRatio?: StandardLonghandProperties['aspectRatio'];
  /**
   * Shorthand for `height`.
   * @note `height` takes precedence over `h` if both are defined.
   */
  h?: StandardLonghandProperties['height'] | number;
  height?: StandardLonghandProperties['height'] | number;
  /**
   * Shorthand for `maxHeight`.
   * @note `maxHeight` takes precedence over `maxH` if both are defined.
   */
  maxH?: StandardLonghandProperties['maxHeight'] | number;
  maxHeight?: StandardLonghandProperties['maxHeight'] | number;
  /**
   * Shorthand for `maxWidth`.
   * @note `maxWidth` takes precedence over `maxW` if both are defined.
   */
  maxW?: StandardLonghandProperties['maxWidth'] | number;
  maxWidth?: StandardLonghandProperties['maxWidth'] | number;
  /**
   * Shorthand for `minHeight`.
   * @note `minHeight` takes precedence over `minH` if both are defined.
   */
  minH?: StandardLonghandProperties['minHeight'] | number;
  minHeight?: StandardLonghandProperties['minHeight'] | number;
  /**
   * Shorthand for `minWidth`.
   * @note `minWidth` takes precedence over `minW` if both are defined.
   */
  minW?: StandardLonghandProperties['minWidth'] | number;
  minWidth?: StandardLonghandProperties['minWidth'] | number;
  /**
   * Shorthand for `width`.
   * @note `width` takes precedence over `w` if both are defined.
   */
  w?: StandardLonghandProperties['width'] | number;
  width?: StandardLonghandProperties['width'] | number;
}

export interface WithDisabled {
  /** @default false */
  disabled?: boolean;
}

export interface WithDisplay {
  display?: StandardLonghandProperties['display'];
}

export interface WithElementSpacing {
  /** @default false */
  prefixSpacing?: boolean | StringOrNumber;
  /** @default false */
  suffixSpacing?: boolean | StringOrNumber;
}

export interface WithEndContent {
  /**
   * Display some content after the main content.
   */
  endContent?: ReactNode;
}

export interface WithFlexBox
  extends Pick<StandardLonghandProperties, 'alignContent' | 'justifyItems'>,
    Pick<StandardShorthandProperties, 'placeContent' | 'placeItems'> {
  /**
   * How to align the contents along the cross axis.
   *
   * Any 'align-items' valid CSS value is accepted.
   */
  align?: StandardLonghandProperties['alignItems'];
  /**
   * How children are placed in the flex container.
   *
   * Any 'flex-direction' valid CSS value is accepted.
   */
  direction?: StandardLonghandProperties['flexDirection'];
  /**
   * The gap CSS property sets the gaps (gutters) between rows and columns.
   * It is a shorthand for row-gap and column-gap.
   */
  gap?: Gap;
  /**
   * How to align the contents along the main axis.
   *
   * Any 'justify-content' valid CSS value is accepted.
   */
  justify?: StandardLonghandProperties['justifyContent'];
  wrap?: StandardLonghandProperties['flexWrap'];
}

export interface WithFlexItem
  extends Pick<StandardLonghandProperties, 'alignSelf' | 'justifySelf' | 'order'> {
  /**
   * A fixed or relative size along its container's main axis.
   */
  basis?: StandardLonghandProperties['flexBasis'];
  /**
   * Set width and/or height to fill the container.
   */
  fill?: boolean | Orientation;
  /**
   * Set flex-grow and/or flex-shrink.
   */
  flex?: boolean | 'grow' | 'shrink' | { grow?: number; shrink?: number };
}

export interface WithFormElements extends WithDisabled {
  name: string;
  /** @default false */
  readOnly?: boolean;
  width?: StringOrNumber;
}

export interface WithGrid {
  area?: StandardShorthandProperties['gridArea'];
  autoColumns?: StandardLonghandProperties['gridAutoColumns'];
  autoFlow?: StandardLonghandProperties['gridAutoFlow'];
  autoRows?: StandardLonghandProperties['gridAutoRows'];
  column?: StandardShorthandProperties['gridColumn'];
  columnEnd?: StandardLonghandProperties['gridColumnEnd'];
  columnGap?: Gap;
  columnStart?: StandardLonghandProperties['gridColumnStart'];
  grid?: StandardShorthandProperties['grid'];
  row?: StandardShorthandProperties['gridRow'];
  rowEnd?: StandardLonghandProperties['gridRowEnd'];
  rowGap?: Gap;
  rowStart?: StandardLonghandProperties['gridRowStart'];
  template?: StandardShorthandProperties['gridTemplate'];
  templateAreas?: StandardLonghandProperties['gridTemplateAreas'];
  templateColumns?: StandardLonghandProperties['gridTemplateColumns'];
  templateRows?: StandardLonghandProperties['gridTemplateRows'];
}

export interface WithHeight {
  /**
   * Component height
   * @default md
   */
  height?: Sizes;
}

export interface WithInline {
  /**
   * Display as an inline element
   * @default false
   */
  inline?: boolean;
}

export interface WithLabel {
  label?: ReactNode;
}

export interface WithLayout extends WithDisplay, WithDimension {
  opacity?: StandardLonghandProperties['opacity'] | number;
  overflow?: StandardShorthandProperties['overflow'];
  pointerEvents?: StandardLonghandProperties['pointerEvents'];
  textAlign?: StandardLonghandProperties['textAlign'];
  transform?: StandardLonghandProperties['transform'];
  transformOrigin?: StandardLonghandProperties['transformOrigin'];
  transition?: StandardShorthandProperties['transition'];
  visibility?: StandardLonghandProperties['visibility'];
}

export interface WithLight {
  /**
   * Remove bold style
   */
  light?: boolean;
}

export interface WithMargin {
  /**
   * Shorthand for `margin`.
   * @note `margin` takes precedence over `m` if both are defined.
   */
  m?: SpacingOrZero;
  /**
   * You can use the shortcuts: m (all sides), mb (bottom), ml (left), mr (right), mt (top), mx (left and right), my (top and bottom)
   */
  margin?: SpacingOrZero;
  /** margin-bottom */
  mb?: SpacingAuto;
  /** margin-left */
  ml?: SpacingAuto;
  /** margin-right */
  mr?: SpacingAuto;
  /** margin-top */
  mt?: SpacingAuto;
  /** margin horizontal axis */
  mx?: SpacingAuto;
  /** margin vertical axis */
  my?: SpacingOrZero;
}

export interface WithOpen {
  /**
   * Controlled status
   */
  open?: boolean;
}

export interface WithOutline
  extends Pick<Theme, 'outlineOffset' | 'outlineOpacity' | 'outlineWidth' | 'outlineZIndex'> {}

export interface WithPadding {
  /**
   * Shorthand for `padding`.
   * @note `padding` takes precedence over `p` if both are defined.
   */
  p?: SpacingOrZero;
  /**
   * You can use the shortcuts: p (all sides), pb (bottom), pl (left), pr (right), pt (top), px (left and right), py (top and bottom)
   */
  padding?: SpacingOrZero;
  /** padding-bottom */
  pb?: SpacingOrZero;
  /** padding-left */
  pl?: SpacingOrZero;
  /** padding-right */
  pr?: SpacingOrZero;
  /** padding-top */
  pt?: SpacingOrZero;
  /** padding horizontal axis */
  px?: SpacingOrZero;
  /** padding vertical axis */
  py?: SpacingOrZero;
}

export interface WithPositioning {
  bottom?: StandardLonghandProperties['bottom'] | number;
  left?: StandardLonghandProperties['left'] | number;
  position?: StandardLonghandProperties['position'];
  right?: StandardLonghandProperties['right'] | number;
  top?: StandardLonghandProperties['top'] | number;
  transform?: StandardLonghandProperties['transform'];
  zIndex?: StandardLonghandProperties['zIndex'] | number;
}

export interface WithRadius {
  radius?:
    | Radius
    | {
        bottom?: Radius;
        left?: Radius;
        right?: Radius;
        top?: Radius;
      };
}

export interface WithShadow {
  shadow?: Shadow | false;
}

export interface WithStartContent {
  /**
   * Display some content before the main content.
   */
  startContent?: ReactNode;
}

export interface WithTextOptions<T extends TextSizes | HeadingSizes = TextSizes>
  extends WithTextSize<T> {
  /**
   * Bold text
   * @default false
   */
  bold?: boolean;
  /**
   * Italic text
   * @default false
   */
  italic?: boolean;
  letterSpacing?: StandardLonghandProperties['letterSpacing'];
  lineHeight?: StandardLonghandProperties['lineHeight'];
  textDecoration?: StandardShorthandProperties['textDecoration'];
  textTransform?: StandardLonghandProperties['textTransform'];
  wordSpacing?: StandardLonghandProperties['wordSpacing'];
}

export interface WithTextSize<T = TextSizes> {
  /**
   * Text size
   */
  size?: T;
}

export interface WithTheme {
  theme: Theme;
}

export interface WithVariant<T extends string = Variant> {
  /**
   * Component variant
   * @default solid
   */
  variant?: T;
}
