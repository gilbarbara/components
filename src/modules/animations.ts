import { keyframes } from '@emotion/react';
import * as scroll from 'scroll';

import * as baseTheme from './theme';

import { Theme, Variants } from '../types';

interface ScrollToOptions {
  element?: HTMLElement;
  scrollDuration?: number;
}

export function animateIcon(
  target: HTMLElement,
  color: Variants = 'primary',
  theme: Theme = baseTheme,
) {
  const { black, colors, gray, white } = theme;

  /* istanbul ignore else */
  if (target) {
    const icon = target.querySelector('[data-component-name="Icon"]') as HTMLSpanElement;

    if (!icon) {
      return;
    }

    const getColor = (variant: Variants) => {
      switch (variant) {
        case 'black': {
          return black;
        }
        case 'white': {
          return white;
        }
        case 'gray': {
          return gray;
        }
        // No default
      }

      return colors[variant];
    };

    const iconClone = document.createElement('span');

    iconClone.innerHTML = icon.innerHTML;
    iconClone.classList.add(icon.className, 'will-animate');
    iconClone.setAttribute(
      'style',
      `color: ${getColor(color)}; position: absolute; top: ${icon.offsetTop}px; left: ${
        icon.offsetLeft
      }px`,
    );
    target.appendChild(iconClone);

    setTimeout(() => {
      iconClone.classList.add('is-animating');
    }, 100);

    target.addEventListener('transitionend', () => {
      /* istanbul ignore else */
      if (iconClone.parentNode !== null && iconClone.classList.contains('is-animating')) {
        iconClone.parentNode.removeChild(iconClone);
      }
    });
  }
}

export const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

export const fadeOut = keyframes`
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
`;

export const fadeInOut = keyframes`
  0% {
    opacity: 0;
  }

  50% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
`;

export const rotate = keyframes`
  100% {
    transform: rotate(360deg);
  }
`;

export function scrollDocument(): HTMLElement {
  return (document.scrollingElement as HTMLElement) || document.createElement('body');
}

export function scrollTo(value: number, options: ScrollToOptions = {}): Promise<void> {
  const { element = scrollDocument(), scrollDuration = 400 } = options;

  return new Promise((resolve, reject) => {
    const { scrollTop } = element;

    const nextValue = scrollDocument().scrollTop + value;
    const limit = nextValue > scrollTop ? nextValue - scrollTop : scrollTop - nextValue;

    scroll.top(
      element,
      nextValue,
      { duration: limit < 100 ? 50 : scrollDuration },
      (error: any) => {
        if (error && error.message !== 'Element already at target scroll position') {
          return reject(error);
        }

        return resolve();
      },
    );
  });
}
