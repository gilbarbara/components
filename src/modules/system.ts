import isPropValid from '@emotion/is-prop-valid';
import { css, CSSObject } from '@emotion/react';
import { capitalize } from '@gilbarbara/helpers';
import { StringOrNumber } from '@gilbarbara/types';
import is from 'is-lite';
import { rgba } from 'polished';

import { getColorVariant, getTheme, px } from './helpers';

import {
  BorderItemSide,
  WithAlign,
  WithBorder,
  WithBorderless,
  WithColor,
  WithDisplay,
  WithElementSpacing,
  WithFlexBox,
  WithFlexItem,
  WithInvert,
  WithLayout,
  WithMargin,
  WithPadding,
  WithPositioning,
  WithRadius,
  WithShadow,
  WithTextColor,
  WithTextOptions,
  WithTheme,
  WithTransparent,
} from '../types';

export function getStyledOptions(...exclude: string[]) {
  return {
    shouldForwardProp: (prop: string) =>
      isPropValid(prop) &&
      ![
        'color',
        'direction',
        'display',
        'height',
        'loading',
        'margin',
        'opacity',
        'overflow',
        'padding',
        'pointerEvents',
        'radius',
        'shadow',
        'size',
        'transform',
        'width',
        'wrap',
        ...exclude,
      ].includes(prop) &&
      !['onClear', 'onCreate', 'onDropdown', 'onOpen', 'onSelect'].some(d => prop.startsWith(d)),
  };
}

export function isDarkMode(props: WithTheme) {
  return !!props?.theme?.darkMode;
}

export function alignStyles<T extends WithAlign>(props: T): CSSObject {
  const { align } = props;

  if (align) {
    return {
      textAlign: align,
    };
  }

  return {};
}

export const appearanceStyles: CSSObject = {
  appearance: 'none',
};

export function backgroundStyles<T extends WithColor & WithInvert & WithTransparent & WithTheme>(
  props: T,
  withBorder = true,
): CSSObject {
  const { variants } = getTheme(props);
  const { invert, shade = 'mid', transparent, variant } = props;

  if (variant) {
    const { bg, color } = getColorVariant(variant, shade, variants);

    const styles: CSSObject = {
      backgroundColor: invert || transparent ? 'transparent' : bg,
      color: invert || transparent ? bg : color,
    };

    if (withBorder) {
      styles.border = transparent ? 0 : `1px solid ${bg}`;
    }

    return styles;
  }

  return {};
}

export function baseStyles<T extends WithColor & WithTheme>(props: T): CSSObject {
  const { fontFamily } = getTheme(props);

  return {
    boxSizing: 'border-box',
    fontFamily,
  };
}

export function borderStyles<T extends WithBorder & WithTheme>(props: T): CSSObject {
  const { border } = props;
  const { variants } = getTheme(props);

  let { bg: borderColor } = getColorVariant('gray', 'lighter', variants);
  const defaultBorder = `1px solid ${borderColor}`;
  let output: CSSObject = {};

  const getBorderValue = (side: BorderItemSide, value: string = defaultBorder) => {
    const item: CSSObject = {};

    if (['bottom', 'left', 'right', 'top'].includes(side)) {
      item[`border${capitalize(side)}`] = value;
    } else if (['start', 'end'].includes(side)) {
      item[`borderInline${capitalize(side)}`] = value;
    } else if (side === 'horizontal') {
      item.borderBottom = value;
      item.borderTop = value;
    } else if (side === 'vertical') {
      item.borderLeft = value;
      item.borderRight = value;
    } else {
      item.border = value;
    }

    return item;
  };

  if (is.nullOrUndefined(border)) {
    return output;
  }

  if (is.boolean(border)) {
    output.border = border ? defaultBorder : undefined;
  } else if (is.string(border)) {
    output = getBorderValue(border);
  } else if (is.array(border)) {
    const items: CSSObject[] = [];

    border.forEach(item => {
      const {
        color,
        shade = 'lighter',
        side,
        size = '1px',
        style = 'solid',
        variant = 'gray',
      } = item;

      ({ bg: borderColor } = getColorVariant(variant, shade, variants));
      const value = `${px(size)} ${style} ${color || borderColor}`;

      items.push(getBorderValue(side, value));
    });

    output = items.reduce<CSSObject>((acc, item) => {
      Object.entries(item).forEach(([key, value]) => {
        acc[key] = value;
      });

      return acc;
    }, {});
  }

  return output;
}

export const buttonStyles: CSSObject = {
  ...appearanceStyles,
  backgroundColor: 'transparent',
  border: 0,
};

export function colorStyles<T extends WithColor & WithTheme>(props: T): CSSObject {
  const { shade = 'mid', variant } = props;
  const { variants } = getTheme(props);

  if (variant) {
    const { bg } = getColorVariant(variant, shade, variants);

    return {
      color: bg,
    };
  }

  return {};
}

export function displayStyles<T extends WithDisplay>(props: T): CSSObject {
  const { display } = props;

  if (display) {
    return { display };
  }

  return {};
}

export function flexBoxStyles<T extends WithFlexBox>(props: T): CSSObject {
  const { align, alignContent, direction, justify, wrap } = props;

  return {
    alignContent,
    alignItems: align,
    flexDirection: direction,
    flexWrap: wrap,
    justifyContent: justify,
  };
}

export function flexItemStyles<T extends WithFlexItem>(props: T): CSSObject {
  const { alignSelf, basis, fill, flex, justifySelf, order } = props;

  const output: CSSObject = {};

  if (!is.nullOrUndefined(fill)) {
    output.height = fill === true || fill === 'vertical' ? '100%' : undefined;
    output.width = fill === true || fill === 'horizontal' ? '100%' : undefined;
  }

  if (!is.nullOrUndefined(flex)) {
    if (is.boolean(flex)) {
      output.flex = flex ? '1 1' : '0 0';
    }

    if (is.string(flex)) {
      output.flex = flex === 'grow' ? '1 0' : '0 1';
    }

    if (is.plainObject(flex)) {
      output.flex = `${flex.grow ? flex.grow : 0} ${flex.shrink ? flex.shrink : 0}`;
    }
  }

  return {
    ...output,
    alignSelf,
    flexBasis: basis,
    justifySelf,
    order,
  };
}

export function inputStyles<
  T extends WithTheme &
    WithBorderless &
    WithElementSpacing & { large?: boolean; multiple?: boolean; width?: StringOrNumber },
>(props: T, type: 'input' | 'select' | 'textarea') {
  const { borderless, large, multiple, prefixSpacing, suffixSpacing, width } = props;
  const darkMode = isDarkMode(props);
  const {
    colors,
    darkColor,
    fontFamily,
    grayDark,
    grayDarker,
    grayLighter,
    grayLightest,
    grayMid,
    inputHeight,
    lightColor,
    radius,
    spacing,
    typography,
    white,
  } = getTheme(props);

  const isSelect = is.boolean(multiple);
  const placeholderColor = grayMid;

  let height;
  let paddingY = large ? spacing.md : spacing.sm;
  let paddingLeft: StringOrNumber | undefined = borderless ? 0 : spacing.md;
  let paddingRight: StringOrNumber | undefined = borderless ? 0 : spacing.md;

  if (type === 'textarea') {
    paddingY = spacing.xs;
  }

  if (type !== 'textarea' && !multiple) {
    height = large ? inputHeight.large : inputHeight.normal;
  }

  if (isSelect) {
    paddingY = large ? spacing.sm : spacing.xs;
    paddingRight = spacing.lg;
  }

  if (prefixSpacing) {
    paddingLeft = is.boolean(prefixSpacing) ? '40px' : px(prefixSpacing);
  }

  if (suffixSpacing) {
    paddingRight = is.boolean(suffixSpacing) ? '40px' : px(suffixSpacing);
  }

  const disabled = css`
    ${!borderless && `background-color: ${darkMode ? grayDark : grayLightest}`};
    border-color: ${darkMode ? grayDark : grayLighter};
    color: ${darkMode ? grayLighter : grayDark};
    cursor: not-allowed;
  `;

  const styles = borderless
    ? css`
        background-color: transparent;
        border: 0;
        border-bottom: 1px solid ${darkMode ? grayDark : grayMid};
      `
    : css`
        background-color: ${darkMode ? grayDarker : white};
        border: 1px solid ${darkMode ? grayDark : grayMid};
        border-radius: ${radius.xs};
      `;

  return css`
    color: ${darkMode ? lightColor : darkColor};
    display: block;
    font-family: ${fontFamily};
    font-size: ${typography.regular.fontSize};
    height: ${height};
    line-height: 1.4;
    padding: ${paddingY} ${paddingRight} ${paddingY} ${paddingLeft};
    width: ${width ? px(width) : '100%'};
    ${styles};

    &:focus {
      ${!!borderless && `border-color: ${colors.primary}`};
      box-shadow: ${borderless ? '' : `0 0 8px 1px ${rgba(colors.primary, 1)}`};
      outline: none;
    }

    &:disabled {
      ${disabled};
    }

    ${!isSelect &&
    css`
      &::placeholder {
        color: ${placeholderColor};
      }

      &:not(:placeholder-shown) {
        border-color: ${colors.primary};
      }

      &:read-only {
        ${disabled};
      }
    `};
  `;
}

export function layoutStyles<T extends WithLayout>(props: T): CSSObject {
  const {
    display,
    height,
    maxHeight,
    maxWidth,
    minHeight,
    minWidth,
    opacity,
    overflow,
    pointerEvents,
    textAlign,
    transform,
    transformOrigin,
    transition,
    width,
  } = props;

  const output: CSSObject = {
    opacity,
    overflow,
    pointerEvents,
    textAlign,
    transition,
    transform,
    transformOrigin,
  };

  if (!is.nullOrUndefined(display)) {
    output.display = display;
  }

  if (!is.nullOrUndefined(height)) {
    output.height = px(height);
  }

  if (!is.nullOrUndefined(maxHeight)) {
    output.maxHeight = px(maxHeight);
  }

  if (!is.nullOrUndefined(maxWidth)) {
    output.maxWidth = px(maxWidth);
  }

  if (!is.nullOrUndefined(minHeight)) {
    output.minHeight = px(minHeight);
  }

  if (!is.nullOrUndefined(minWidth)) {
    output.minWidth = px(minWidth);
  }

  if (!is.nullOrUndefined(width)) {
    output.width = px(width);
  }

  return output;
}

export function marginStyles<T extends WithMargin & WithTheme>(props: T): CSSObject {
  const { margin, mb, ml, mr, mt, mx, my } = props;
  const { spacing } = getTheme(props);

  const output: CSSObject = {};

  if (!is.undefined(margin)) {
    output.margin = margin ? spacing[margin] : margin;
  }

  if (!is.undefined(my)) {
    output.marginBottom = my ? spacing[my] : my;
    output.marginTop = my ? spacing[my] : my;
  } else {
    if (!is.undefined(mb)) {
      output.marginBottom = mb === 'auto' || mb === 0 ? mb : spacing[mb];
    }

    if (!is.undefined(mt)) {
      output.marginTop = mt === 'auto' || mt === 0 ? mt : spacing[mt];
    }
  }

  if (!is.undefined(mx)) {
    output.marginLeft = mx === 'auto' || mx === 0 ? mx : spacing[mx];
    output.marginRight = mx === 'auto' || mx === 0 ? mx : spacing[mx];
  } else {
    if (!is.undefined(ml)) {
      output.marginLeft = ml === 'auto' || ml === 0 ? ml : spacing[ml];
    }

    if (!is.undefined(mr)) {
      output.marginRight = mr === 'auto' || mr === 0 ? mr : spacing[mr];
    }
  }

  return output;
}

export function outlineStyles<T extends WithTheme>(props: T): CSSObject {
  const { colors } = getTheme(props);

  return {
    boxShadow: `0 0 6px 0 ${rgba(colors.primary, 0.6)}`,
  };
}

export function paddingStyles<T extends WithPadding>(props: T, force = false): CSSObject {
  const { padding, pb, pl, pr, pt, px: paddingX, py } = props;
  const { spacing } = getTheme(props);

  const output: CSSObject = {};

  if (padding) {
    output.padding = spacing[padding];
  }

  if (py) {
    output.paddingBottom = spacing[py];
    output.paddingTop = spacing[py];
  } else {
    if (pb) {
      output.paddingBottom = spacing[pb];
    }

    if (pt) {
      output.paddingTop = spacing[pt];
    }
  }

  if (paddingX) {
    output.paddingLeft = spacing[paddingX];
    output.paddingRight = spacing[paddingX];
  } else {
    if (pl) {
      output.paddingLeft = spacing[pl];
    }

    if (pr) {
      output.paddingRight = spacing[pr];
    }
  }

  if (force) {
    return Object.entries(output).reduce<CSSObject>((acc, [key, value]) => {
      acc[key] = `${value} !important`;

      return acc;
    }, {});
  }

  return output;
}

export function positioningStyles<T extends WithPositioning>(props: T): CSSObject {
  const { bottom, left, position, right, top, zIndex } = props;

  const output: CSSObject = {
    position,
    zIndex,
  };

  if (!is.nullOrUndefined(bottom)) {
    output.bottom = px(bottom);
  }

  if (!is.nullOrUndefined(left)) {
    output.left = px(left);
  }

  if (!is.nullOrUndefined(right)) {
    output.right = px(right);
  }

  if (!is.nullOrUndefined(top)) {
    output.top = px(top);
  }

  return output;
}

export function radiusStyles<T extends WithRadius & WithTheme>(props: T): CSSObject {
  const { radius } = getTheme(props);

  if (props.radius) {
    return {
      borderRadius: radius[props.radius],
    };
  }

  return {};
}

export function shadowStyles<T extends WithShadow & WithTheme>(props: T): CSSObject {
  const { shadow } = getTheme(props);

  if (props.shadow) {
    return {
      boxShadow: isDarkMode(props)
        ? shadow[props.shadow].replace(/148/g, '222')
        : shadow[props.shadow],
    };
  }

  return {};
}

export function textColorStyles<T extends WithTextColor & WithTheme>(props: T): CSSObject {
  const { color, colorShade = 'mid', colorVariant } = props;
  const { variants } = getTheme(props);

  if (color) {
    return {
      color,
    };
  }

  if (colorVariant) {
    const { bg } = getColorVariant(colorVariant, colorShade, variants);

    return {
      color: bg,
    };
  }

  return {};
}

export function textStyles<T extends WithTextOptions & WithTheme>(
  props: T,
  lineHeightCustom?: StringOrNumber,
): CSSObject {
  const { bold = false, size } = props;
  const { typography } = getTheme(props);

  if (size) {
    const { fontSize, lineHeight, weight } = typography[size];
    const fontWeight = bold ? weight[1] : weight[0];

    return {
      fontSize,
      fontWeight,
      lineHeight: lineHeightCustom || lineHeight,
    };
  }

  return {};
}
