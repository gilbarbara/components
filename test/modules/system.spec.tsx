import { alignStyles, getContainerStyles, getStyledOptions, isDarkMode } from 'modules/system';

describe('getContainerStyles', () => {
  it('should return properly', () => {
    const { styles } = getContainerStyles({});

    expect(styles).toMatchSnapshot();
  });

  it('should return properly with "responsive: false"', () => {
    const { styles } = getContainerStyles({}, { responsive: false });

    expect(styles).toMatchSnapshot();
  });

  it('should return properly with "verticalPadding: true"', () => {
    const { styles } = getContainerStyles({}, { verticalPadding: true });

    expect(styles).toMatchSnapshot();
  });
});

describe('getStyledOptions', () => {
  it.each([
    ['onChange', true],
    ['color', false],
  ])('should forward %p: %p', (prop, expected) => {
    const { shouldForwardProp } = getStyledOptions();

    expect(shouldForwardProp(prop)).toBe(expected);
  });
});

describe('isDarkMode', () => {
  it('should return the default', () => {
    expect(isDarkMode({})).toBe(false);
  });

  it('should return true', () => {
    expect(isDarkMode({ theme: { darkMode: true } })).toBe(true);
  });
});

describe('alignStyles', () => {
  it('should return properly', () => {
    expect(alignStyles({ align: 'center' })).toMatchSnapshot();
  });

  it('should return an empty object without the "align" prop', () => {
    expect(alignStyles({})).toMatchSnapshot();
  });
});
