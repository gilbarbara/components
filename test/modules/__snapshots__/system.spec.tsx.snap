// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alignStyles should return an empty object without the "align" prop 1`] = `{}`;

exports[`alignStyles should return properly 1`] = `
{
  "textAlign": "center",
}
`;

exports[`getContainerStyles should return properly 1`] = `
"
    padding-left: 16px;
    padding-right: 16px;

    

    @media screen and (min-width: 1024px){padding-left:32px;padding-right:32px;};

    
  "
`;

exports[`getContainerStyles should return properly with "responsive: false" 1`] = `
"
    padding-left: 16px;
    padding-right: 16px;

    

    

    
  "
`;

exports[`getContainerStyles should return properly with "verticalPadding: true" 1`] = `
"
    padding-left: 16px;
    padding-right: 16px;

    
      padding-bottom: 16px;
      padding-top: 16px;
    ;

    @media screen and (min-width: 1024px){padding-left:32px;padding-right:32px;};

    @media screen and (min-width: 1024px){padding-bottom:32px;padding-top:32px;};
  "
`;
